src/meshes/t8code_mesh.jl

```diff
@@ -885,33 +886,79 @@ function adapt(forest::Ptr{t8_forest}, adapt_callback; recursive = true, balance
+                    if an element should be refined, coarsened or stay unchanged.
+
+    The expected callback signature is as follows:
+
+      `adapt_callback(forest, ltreeid, eclass_scheme, lelemntid, elements, is_family, user_data)`
+        # Arguments
+        - `forest`: Pointer to the analyzed forest.
+        - `ltreeid`: Local index of the current tree where the analyzed elements are part of.
+        - `eclass_scheme`: Element class of `elements`.
+        - `lelemntid`: Local index of the first element in `elements`.
+        - `elements`: Array of elements. If consecutive elements form a family
+                      they are passed together, otherwise `elements` consists of just one element.
+        - `is_family`: Boolean signifying if `elements` represents a family or not.
+        - `user_data`: Void pointer to some arbitrary user data. Default value is `C_NULL`.
+        # Returns
+          -1 : Coarsen family of elements.
+           0 : Stay unchanged.
+           1 : Refine element.
+
+- `kwargs`:
+    - `recursive = true`: Adapt the forest recursively. If true the caller must ensure that the callback
+                          returns 0 for every analyzed element at some point to stop the recursion.
+    - `balance = true`: Make sure the adapted forest is 2^(NDIMS-1):1 balanced.
+    - `partition = true`: Partition the forest to redistribute elements evenly among MPI ranks.
+    - `ghost = true`: Create a ghost layer for MPI data exchange.
+    - `user_data = C_NULL`: Pointer to some arbitrary user-defined data.
+
+Returrns nothing.
+"""
 function adapt!(mesh::T8codeMesh, adapt_callback; kwargs...)
     # Call `t8_forest_ref(Ref(mesh.forest))` to keep it.
-    mesh.forest = adapt(mesh.forest, adapt_callback; kwargs...)
+    update_forest!(mesh, adapt_forest(mesh.forest.pointer, adapt_callback; kwargs...))
+    return nothing
 end
 
-"""
-    Trixi.balance!(mesh::T8codeMesh)
-
-Balance a `T8codeMesh` to ensure 2^(NDIMS-1):1 face neighbors.
-"""
-function balance!(mesh::T8codeMesh)
+function balance_forest(forest::Ptr{t8_forest})
     new_forest_ref = Ref{t8_forest_t}()
     t8_forest_init(new_forest_ref)
     new_forest = new_forest_ref[]
 
-    let set_from = mesh.forest, no_repartition = 1, do_ghost = 1
+    let set_from = forest, no_repartition = 1, do_ghost = 1
         t8_forest_set_balance(new_forest, set_from, no_repartition)
         t8_forest_set_ghost(new_forest, do_ghost, T8_GHOST_FACES)
         t8_forest_commit(new_forest)
     end
 
-    mesh.forest = new_forest
+    return new_forest
+end
 
+"""
+    Trixi.balance!(mesh::T8codeMesh)
+
+Balance a `T8codeMesh` to ensure 2^(NDIMS-1):1 face neighbors.
+
+# Arguments
+- `mesh::T8codeMesh`: Initialized mesh object.
+
+Returns noothing.
+"""
+function balance!(mesh::T8codeMesh)
+    update_forest!(mesh, balance_forest(mesh.forest.pointer))
     return nothing
 end
 
-function partition(forest::Ptr{t8_forest})
+function partition_forest(forest::Ptr{t8_forest})
```

Comment:
```suggestion
function partition_forest(forest::Union{T8code.ForestWrapper, Ptr{t8_forest}})
```
See for `balance_forest`.